ui/app/src/model/datasource-api.ts

```diff
@@ -11,45 +11,206 @@
 
-export function useDatasourceApi(): DatasourceApi {
-  return {
-    getDatasource: async (project, selector) => {
-      return fetchDatasourceList(project, selector.kind, selector.name ? undefined : true, selector.name).then(
-        (list) => {
-          // hopefully it should return at most one element
-          if (list[0] !== undefined) {
-            return {
-              resource: list[0],
-              proxyUrl: getProxyUrl(list[0]),
-            };
-          }
-        }
-      );
-    },
-    getGlobalDatasource: async (selector) => {
-      return fetchGlobalDatasourceList(selector.kind, selector.name ? undefined : true, selector.name).then((list) => {
-        // hopefully it should return at most one element
-        if (list[0] !== undefined) {
-          return {
-            resource: list[0],
-            proxyUrl: getProxyUrl(list[0]),
-          };
-        }
-      });
-    },
-
-    listDatasources: async (project, pluginKind) => {
-      return fetchDatasourceList(project, pluginKind);
-    },
-
-    listGlobalDatasources: async (pluginKind) => {
-      return fetchGlobalDatasourceList(pluginKind);
-    },
-  };
+export class HTTPDatasourceAPI implements DatasourceApi {
+  getDatasource(
+    project: string,
+    selector: DatasourceSelector
+  ): Promise<{ resource: Datasource; proxyUrl: string } | undefined> {
+    return fetchDatasourceList(project, selector.kind, selector.name ? undefined : true, selector.name).then((list) => {
+      // hopefully it should return at most one element
+      if (list[0] !== undefined) {
+        return {
+          resource: list[0],
+          proxyUrl: getProxyUrl(list[0]),
+        };
+      }
+    });
+  }
+
+  getGlobalDatasource(
+    selector: DatasourceSelector
+  ): Promise<{ resource: GlobalDatasource; proxyUrl: string } | undefined> {
+    return fetchGlobalDatasourceList(selector.kind, selector.name ? undefined : true, selector.name).then((list) => {
+      // hopefully it should return at most one element
+      if (list[0] !== undefined) {
+        return {
+          resource: list[0],
+          proxyUrl: getProxyUrl(list[0]),
+        };
+      }
+    });
+  }
+
+  listDatasources(project: string, pluginKind?: string): Promise<Datasource[]> {
+    return fetchDatasourceList(project, pluginKind);
+  }
+
+  listGlobalDatasources(pluginKind?: string): Promise<GlobalDatasource[]> {
+    return fetchGlobalDatasourceList(pluginKind);
+  }
+}
+
+class Cache {
+  private datasources: LRUCache<string, Datasource>;
+  private emptyDatasources: LRUCache<string, boolean>;
+  private globalDatasources: LRUCache<string, GlobalDatasource>;
+
+  constructor() {
+    // We want to have a cache expiration set to 5min and removed automatically the old values.
+    // The option below is setting that.
+    // Note: TTL (Time To Leave) is in millisecond.
+    const option = { ttl: 5 * 60 * 1000, ttlAutopurge: true };
```

Comment:
Are there any weird edge cases someone could end up in if these values are cached for 5 minutes and something changes in the backend? I'm guessing it's fine, but I have a little less context on this bit of the code, so wanted to double-check.

Reply:
Yeah that should be fine. The datasource configuration shouldn't change that often.
And if it's the case then simply reloading the page will refresh the cache. That's not so hurtful I think.